### DISCUSSION TITLE
What optimizations does GraphQL afford?

### DISCUSSION THREAD

**Discussion by abaco** [gatsbyjs/gatsby#27894] [Help] [answered]
I'm reviewing Gatsby for my employer, and I have a question that I can't reply on my own. I've read all I could find about GraphQL in the docs, as well as issues #420 and and #4994, and found this [remark](https://github.com/gatsbyjs/gatsby/issues/4994#issuecomment-382110077) from @KyleAMathews:

> I wanted a way for users to *declaratively* specify (aka write a query) for what data a page component needs as there's a ton of optimizations we can do because of this

My question is: what optimizations have been implemented / planned, and in general what kind of optimizations are possible?

**Comment by sidharthachatterjee** [comment]
@abaco Good question. One of the biggest benefits of GraphQL is the fact that it prevents both under-fetching and over-fetching. What this means is that the data that is present in your site's bundle ends up only being the keys you specify in your GraphQL queries (which you use in your components). This prevents the common issue where we tend to load more data than what is used in web projects, leading to a slowdown in load times etc

The other advantage (not an optimisation per se) that GraphQL affords in Gatsby specifically is that it enables fetching data from multiple services or data formats (via plugins) in a conveniently coherent way. You don't need to write API calls for respective APIs of services for data that you'd like to use at build time. They are all available (via gatsby and its plugins) in GraphQL. This makes access to data simpler as well. 

Hope this helps!

**Reply by abaco** [reply]
@sidharthachatterjee Thanks for your reply. I'm afraid my question wasn't clear, though. I'm really asking the same question @jamesplease asked in #4994: why the GraphQL query language was preferred over the default option of just using plain JavaScript? I couldn't find any explanation in the docs. What are usually presented as advantages of GraphQL are rather advantages of a separate data layer: the ability to collect and combine data from multiple sources, and to query it from components in a coherent way, selecting only what's needed. This is clear to me. But why use GraphQL to query the local database?

Kyle's answer in #4994 is good, if only partial. He said it was important to use a *declarative* language, and that requirement alone excludes JavaScript. Why declarative? Because then you can do optimizations. So, again, what kind of optimizations are we talking about? What optimization does a declarative query language affords us?

Given the requirement of a declarative language, GraphQL is arguably a good choice, if not the best. But why choose a declarative language over the more expressive JavaScript?

**Reply by herecydev** [reply]
> But why use GraphQL to query the local database

To prevent overfetching as Sidhartha raised; with GraphQL you can only receive *exactly* what you asked for. With regular javascript objects, they can come bloated with additional properties as the complexity of the application/data increases. It's very difficult to avoid in a maintanable way, either you accept a single model with lots of properties or you have lots of bespoke models. GraphQL is really solving that problem by being a querying language.

**Reply by abaco** [reply]
I think I should be more concrete, but I don't want to trivialize this matter. Let me try.

At an abstract level, there's no question JavaScript is more expressive than GraphQL. Speaking more concretely, selecting a subset of properties from a JS object should be as simple as doing `_.pick(obj, ['prop1', 'prop2'])`. Since the database is local, one could coarsely select (rather than "fetch") a subset of the whole data, and then traverse it even multiple times to refine the selection, apply transformations, etc. Common cases should be handled by convenience functions, while for more specific cases the full power of JavaScript would be available. I'm not suggesting designing such an API would be easy, but it's definitely doable. And in the worst case it would be inconvenient, while with GraphQL it's not that hard to hit roadblocks. Say for some reason you need the first and last elements of an array: you can't express that in GraphQL.

Instead of exporting a GraphQL query, components could export a function that operates on the local database and returns the desired data. Gatsby would then inject that data as props in the component, just like it does now with the GraphQL query result.

By the way, I'm not proposing Gatsby steer in this direction. I just understood there were additional motivations for using GraphQL, because Kyle's quote above clearly states so. If instead you say preventing overfetching is the official motivation, that's also fine for me. I just want to know the rationale behind the decision.

**Reply by herecydev** [reply]
> Say for some reason you need the first and last elements of an array: you can't express that in GraphQL.

I don't think thats true. GraphQL doesn't do much out the box but it allows for custom directives to be added, so you could easily build one that resolves the first and last elements. Granted, it's not infinitely flexible but it's often simple enough where it needs to be

I think the motiviation is to have a declarative, well-understood way of data-fetching. Designing a custom API like you've suggested is possible, but it presents yet another barrier for learning a framework. I'm not sure what optimizations graphQL is specficially affording though

**Reply by abaco** [reply]
I think the Gatsby team should take a hard look at the role of GraphQL within the framework. I’ve spent a lot of time examining this matter and, frankly, I came to the conclusion that adopting GraphQL was a mistake.

As users we aren’t provided with a well articulated rationale for why GraphQL was introduced. Instead, the documentation conflates GraphQL with the data layer, listing as advantages of the former what are actually features of the latter. Elsewhere we’re explained that opting out of GraphQL prevents us from using many other useful features of Gatsby, but this just tells how Gatsby is structured, not why. As a result we aren’t able to judge whether the use of GraphQL involves a trade-off that’s good for us.

One should think twice before replacing JavaScript with a special-purpose language like GraphQL, but you don’t seem to have really thought through it *once*. The argument that GraphQL is a well understood way of querying data, almost a de-facto standard, is weak because GraphQL was created to minimize the number of requests and their payloads when querying a *remote* database, while Gatsby’s database is local. The documentation even [touts](https://www.gatsbyjs.com/docs/graphql-concepts/#how-do-graphql-and-gatsby-work-together) the *uniqueness* of Gatsby’s approach to querying a local db, implicitly admitting that for that use case the default choice is a different one - namely, JavaScript. However the JavaScript route was apparently never considered.

As for the optimizations that are possible with GraphQL, I think they were never implemented. And that’s perfectly fine, because there’s really no point in optimizing queries against a local, in memory database, not to mention that they’re only run at build time.
 
Gatsby users routinely run into issues connected to the use of GraphQL. The question is: is it worth it? If there are specific advantages to the GraphQL language compared to JavaScript they should be discussed in the documentation, so that beginners would be better able to understand what they’re signing up to. If instead the advantages are not so clear it may be worth to add a JavaScript API for querying the local database alongside the current GraphQL interface. Source plugins wouldn’t need to be touched, while transformer plugins would probably require some adaptations.

**Reply by herecydev** [reply]
> because GraphQL was created to minimize the number of requests and their payloads when querying a remote database, while Gatsby’s database is local

That's true the advantage of minimising requests isn't relevant. Whether you use graphql or you pass a JS object, a single request will still be made during hydration. However the advantage of minimising the size of the request is there. Consider the following:

```
createPage({
  path: "/",
  component: someComponent
  context: {
    massiveArray: [],
    bigObject: {}
  },
})
```

If the object was massive, Gatsby has no choice but to request all of this as props (as it has no understanding what's used and what isn't). In the case of graphQL it's baked into the design of the language to prevent this happening.

> One should think twice before replacing JavaScript with a special-purpose language like GraphQL

Nobodys forcing this, you can happily pass pure javascript objects if you feel that's best. You can also fetch data at runtime as well. Gatsby has a number of tools to make static site rendering leaner and efficient but you don't have to use them if it's not appropriate.

> If instead the advantages are not so clear it may be worth to add a JavaScript API for querying the local database alongside the current GraphQL interface

Maybe it's worth putting an RFC together if you feel strongly enough that it delivers value. In any case, relevant to the comment above you could do this already (even if Gatsby doesn't support this). You could put together a plugin that uses `onCreatePage` and then calls this new "javascript api" and pulls out the right data for the page.

**Reply by abaco** [reply]
@herecydev You're making the same comparison as in [Gatsby without GraphQL](https://www.gatsbyjs.com/docs/using-gatsby-without-graphql/), namely data-layer-cum-GraphQL vs. no data layer. I'm aware of the positive sides of the data layer, from the many source plugins to live-reloading, but I don't see them as necessarily linked to GraphQL. In short, I'd like to use the data layer without GraphQL.

When did *GraphQL* become shorthand for *data layer*? That's not helpful. For instance that article should be titled *Gatsby without the Data Layer*, because that's actually what it discusses. This conceptual confusion hides the fact that GraphQL is logically independent from the rest of Gatsby's architecture. You could have source plugins without GraphQL. You could have live-reloading without GraphQL. You could prevent over-fetching without GraphQL. Whether it would be *better* to do all those things without GraphQL is a different question, but one that deserves to be considered carefully.

For the sake of the argument I alluded at the possibility to query data through a JavaScript API. My very rough idea is to replace a page query like:

```js
export const query = graphql`
  query {
    site {
      siteMetadata {
        title
      }
    }
  }
`
```

with a JavaScript function like:

```js
export const query = (store):  => {
  return selectFromStore(store, {
    site: select({
      siteMetadata: select('title')
    })
  });
};
```

where `select` returns an operation on a node (i.e. a special object), and `selectFromStore` returns a plain object with the result of the query.

This would enable nested queries and data transformations like so:

```js
import { blogQuery } from '../components/Blog';
import { myTransformation } from '../data-processing';

// ...
// Page component here
// ...

export const query = (store) => {
  const results = selectFromStore(store, {
    site: select({
      siteMetadata: select('title')
        .transform((t) => t.toUppercase()), // transform single field
      blogContent: blogQuery('content') // imported query instead of fragment
    })
  });

  // transform whole tree
  const resultsWithFluidImages = processDataSharp(sharpConfig, results); 
  const resultsWithCustomTransformation = myTransformation(resultsWithFluidImages);

  return resultsWithCustomTransformation;
};
```

This is admittedly very fuzzy. I also doubt I'll make an RFC any time soon, because I don't know much about Gatsby's internals.

In any case I'd like this discussion to remain at the abstract level, not my specific needs or the specific alternative I just quickly threw together.

**Reply by herecydev** [reply]
What you're suggesting above is reasonable but is tied *entirely* to Gatsby. So it presents another barrier to learning. If you have a solution that solves:

- Easy to learn (preferably already learnt outside of Gatsby)
- Performant
- Prevents over-fetching (in turn performance, seo, etc)
- ....

Then definitely create an RFC. Right now it seems that you're suggesting a few Gatsby specific helper functions that can leveraged, but that's more stuff that people will have to learn and reason with. See [here](https://github.com/gatsbyjs/gatsby/discussions/27856) why that's a big problem.

**Comment by vladar** [comment, marked as answer]
> and in general what kind of optimizations are possible?

Some examples of what's possible:

1. Declarative data requirements may help us to eventually switch to lazy sourcing mode (for develop and per-page builds). By analyzing your page query we can figure out which data this page needs before we actually have this data available in the local storage. Then we can ask source plugins to `source` data required for this page only (and postpone sourcing everything else).

2. Someday we may have Gatsby GraphQL available as a remote service as well. In this case, all the network-related benefits of GraphQL will be also important.

3. We can hook external GraphQL APIs into Gatsby schema and provide a unified querying interface for those. [We do this already](https://www.gatsbyjs.com/docs/third-party-graphql/) with `gatsby-source-graphql` but at the moment it acts like a proxy at the query step (not the source nodes step).

4. Declarative queries enable efficient "live" queries. Say, a new node is sourced - you can match this node against registered queries and just update results of the matching queries without re-running all of those. Could be useful for near real-time incremental re-builds. This is somewhat similar to [KSQL](https://hevodata.com/learn/kafka-ksql-streaming-sql-for-kafka/) (but there are also other similar projects).

This is just what immediately came to mind. Also, to clarify: I partially agree with your attitude on this topic and currently we don't utilize the power of declarative queries and schema to its full capacity.

But also as @herecydev mentioned - JavaScript API for data-querying will likely be unique to Gatsby. So it won't help much with the learning curve + this learning will be only useful for Gatsby projects.

**Reply by abaco** [reply]
Thanks @vladar, that's really helpful! Now I can see how GraphQL fits in your vision for Gatsby.

I hope I didn't upset anybody, and that this information will be useful to other Gatsby users and contributors.
